### ISSUE TITLE
OpenTelemetry doesn't seem to work on Bun 0.7.0

### ISSUE THREAD

**Issue by lewisl9029** [oven-sh/bun#3775]
### What version of Bun is running?

0.7.0

### What platform is your computer?

Darwin 22.4.0 arm64 arm

### What steps can reproduce the bug?

I've set up a repo to repro here with instructions: https://github.com/lewisl9029/bun-opentelemetry-issue-repro

Pasting here for convenience:

1. Run `bun run test-server`
2. Run `bun run bun` in a separate terminal
3. Go to http://localhost:58050 in a browser

### What is the expected behavior?

- Request to /v1/traces shows up in test-server terminal

### What do you see instead?

- Nothing shows up

### Additional information

Also provided a node server to demonstrate expected behavior, just run `npm run node` in step 2.

If I had to guess, the fact that bun lacks support for node's --experimental-loader=@opentelemetry/instrumentation/hook.mjs (or the --require=./app/tracing.cjs equivalent for setting up OTel in node with CJS) might be part of the problem? 

Or it could also be something else entirely. OTel's tries to fail silently without affecting program behavior as much as possible, which means we'd probably have to dig down a lot deeper to find the root cause of something like this.

**Comment by birkskyum** [comment]
I'm not able to get the terminal output with `npm run node`.

**Comment by nicu-chiciuc** [comment]
This is the only thing that stops us from migrating from Node to Bun. There aren't clear instruction for how to add telemetry to Bun.
I'm not entirely sure if this is the responsibility of this repo, or, if it should be added to `opentelemetry-js` https://github.com/open-telemetry/opentelemetry-js

**Comment by jonataswalker** [comment]
This is indeed a no-go for some of us. [They are](https://github.com/open-telemetry/opentelemetry-js/issues/2293) struggling to provide support for Deno.

**Comment by zx8086** [comment]
This is one of two things stopping us committing to Bun over NodeJS, Opentelemetry instrumentation and connecting to a Couchbase database

**Comment by kytay** [comment]
Hi there. 

I am trying to following the example on opentelemetry.js on how to setup auto instrumentation using `--require instrumentation.ts` in the command line.

Is using `--require` a must? Is it the reason why there isn't any documentation or articles online that explains how can we use opentelemetry.js with bun?

If I have only an index.ts

```typescript
import express, { Express } from 'express';
import { NodeSDK } from '@opentelemetry/sdk-node';
import { ConsoleSpanExporter } from '@opentelemetry/sdk-trace-node';
import { getNodeAutoInstrumentations } from '@opentelemetry/auto-instrumentations-node';
import {
  PeriodicExportingMetricReader,
  ConsoleMetricExporter,
} from '@opentelemetry/sdk-metrics';

const sdk = new NodeSDK({
  traceExporter: new ConsoleSpanExporter(),
  metricReader: new PeriodicExportingMetricReader({
    exporter: new ConsoleMetricExporter(),
  }),
  instrumentations: [getNodeAutoInstrumentations()],
});

sdk.start();

const PORT: number = parseInt(process.env.PORT || '8080');
const app: Express = express();

function getRandomNumber(min: number, max: number) {
  return Math.floor(Math.random() * (max - min) + min);
}

app.get('/rolldice', (req, res) => {
  res.send(getRandomNumber(1, 6).toString());
});

app.listen(PORT, () => {
  console.log(`Listening for requests on http://localhost:${PORT}`);
});
```

why this code won't work? I did `sdk.start()` before my application code, as per described in [opentelemetry](https://opentelemetry.io/docs/instrumentation/js/getting-started/nodejs/#run-the-instrumented-app)

**Comment by ImLunaHey** [comment]
the code needs to run before any of the imports. that's why it needs the `--require`

**Comment by AlexandrePh** [comment]
does bun offer the require functionality?
Any updates on this issue?

**Comment by ImLunaHey** [comment]
Here is a working example for Bun. Thank you @Jarred-Sumner (https://discord.com/channels/876711213126520882/876711213126520885/1196585812486279208)

```ts
/*instrumentation.ts*/
// Require dependencies
import { NodeSDK } from '@opentelemetry/sdk-node';
import { ConsoleSpanExporter } from '@opentelemetry/sdk-trace-node';
import { getNodeAutoInstrumentations } from '@opentelemetry/auto-instrumentations-node';
import { PeriodicExportingMetricReader, ConsoleMetricExporter } from '@opentelemetry/sdk-metrics';

const sdk = new NodeSDK({
  traceExporter: new ConsoleSpanExporter(),
  metricReader: new PeriodicExportingMetricReader({
    exporter: new ConsoleMetricExporter(),
  }),
  instrumentations: [getNodeAutoInstrumentations()],
});

sdk.start();

```

```ts
/*app.ts*/
import express from 'express';

const PORT = parseInt(process.env.PORT || '8080');
const app = express();

function getRandomNumber(min: number, max: number) {
  return Math.floor(Math.random() * (max - min) + min);
}

app.get('/rolldice', (req, res) => {
  res.send(getRandomNumber(1, 6).toString());
});

app.listen(PORT, () => {
  console.log(`Listening for requests on http://localhost:${PORT}`);
});
```

To run the example: 
```bash
bun run --preload ./instrumentation.ts  ./app.ts
```

**Comment by chlorophant** [comment]
Did this end up working as expected?

**Comment by ImLunaHey** [comment]
Yes. The only thing you need to do differently is use the `preload` flag instead of `require`.

**Comment by ricardo-devis-agullo** [comment]
This solution kinda worked for me, but at least when using `applicationinsights` package (which uses opentelemetry internally), auto collection of http requests still didn't work (on Elysia server).

**Comment by ricardo-devis-agullo** [comment]
I got a minimal reproduction example that doesn't even need Elyisia, just Bun.serve

```javascript
import appInsights from "applicationinsights";
import http from "http";

appInsights.setup("YOUR-CONNECTION-STRING").start();

appInsights.defaultClient.addTelemetryProcessor((envelope) => {
  envelope.data.baseType === "RequestData" && console.log("GOT REQUEST DATA");
});

const PORT = 3000;

function bunServer() {
  Bun.serve({
    port: PORT,
    fetch(req) {
      return new Response("Hello World\n");
    },
  });
}

function nodeServer() {
  const server = http.createServer((req, res) => {
    res.writeHead(200, { "Content-Type": "text/plain" });
    res.end("Hello World\n");
  });
  server.listen(PORT);
}

// This will output "GOT REQUEST DATA" after 5 seconds
const app = nodeServer;
// This will never output "GOT REQUEST DATA"
// const app = bunServer;

app();

setTimeout(() => {
  fetch(`http://localhost:${PORT}`);
}, 5000);

```

**Comment by zx8086** [comment]
> ```shell
> bun run --preload ./instrumentation.ts  ./app.ts
> ```

This solves it

**Comment by ricardo-devis-agullo** [comment]
The thing I mentioned it doesn't, because I think it boils down to appInsights trying to attach itself to `http` module to listen to requests, and Bun.serve just using something completely different.

**Comment by zx8086** [comment]
> Did this end up working as expected?

What didn't work for you ?

**Comment by dusty-phillips** [comment]
I experienced similar to ricardo-devis-agullo. I am able to send manual traces using a variation of ImLunaHey's code, but autoinstrumenting the http endpoints doesn't work.

I have a feeling that the autoinstrumentation for express works differently from the instrumentation for http, so ImLunaHey's code works with express but not Elysia or (in my case) Hono.

My workaround is to manually create spans in a tracing middleware. Because it took me a bloody long time to track down all the attributes, I'm sharing my (Hono-style, but easily adapted) middleware for other users:

```ts
import { createMiddleware } from "hono/factory"
import opentelemetry from "@opentelemetry/api"

const tracer = opentelemetry.trace.getTracer("<your-app-name>")

export const tracingMiddleware = createMiddleware(async (c, next) => {
  await tracer.startActiveSpan(c.req.path, async (span) => {

    // TODO: wasn't able to find http.flavour (http version number)
    span.setAttribute("http.route", c.req.path)
    span.setAttribute("http.method", c.req.method)
    span.setAttribute("http.url", c.req.url)
    span.setAttribute("http.schema", new URL(c.req.raw.url).protocol)
    span.setAttribute("http.referrer", c.req.raw.referrer)
    span.setAttribute("http.client_ip", c.env.requestIP(c.req.raw))
    span.setAttribute("http.user_agent", c.req.header("User-Agent") ?? "")
    span.setAttribute(
      "http.request_content_length",
      c.req.header("Content-Length") ?? "",
    )
    span.setAttribute("span.kind", "server")
    span.setAttribute("trace.parent_id", c.req.header("traceparent") ?? "")
    span.setAttribute("type", "server")

    await next()

    span.setAttribute("http.status_code", c.res.status)
    span.setAttribute("status_code", c.res.status)
    // TODO: would like to get response content length in here

    span.end()
  })
})
```

**Comment by ImLunaHey** [comment]
you could have used this for fetch calls https://github.com/open-telemetry/opentelemetry-js-contrib/tree/main/plugins/node/instrumentation-undici/

**Comment by SlexAxton** [comment]
yea, i think I'm experiencing the same thing:

```
@opentelemetry/instrumentation-http Module http has been loaded before @opentelemetry/instrumentation-http so it might not work, please initialize it before requiring http
@opentelemetry/instrumentation-http Module https has been loaded before @opentelemetry/instrumentation-http so it might not work, please initialize it before requiring https
```

I tried explicitly importing this module into my `telemetry.ts` file, but to no effect.

**Comment by setyongr** [comment]
somehow instrumentation does not work with import statement but works with require

Not Work
```js
import { pino } from 'pino'
export const logger = pino()
```

Work
```js
import type { Logger } from 'pino';
const pino = require('pino');
export const logger: Logger = pino();
```

**Comment by sculpt0r** [comment]
Based on https://github.com/open-telemetry/opentelemetry-js/issues/5367#issuecomment-2611990770 it has to do with `http` module 🤔

**Comment by SirMatGuy** [comment]
Yes, i did some tests, and bun's `http` module emits the `close` event before `end`, which cause the error being shown, although it looks like it can be ignored, the collector still receives the data

**Comment by theshadow27** [comment]
I took a stab at this. [#23798](https://github.com/oven-sh/bun/pull/23798) adds a native zig API on `Bun.telemetry` with minimal surface area - basically integer ID tracking - so that stuff from `'@opentelemetry/sdk-trace-node'` packages have something to attach spans to. 

Works for both `Bun.serve()`  and `http.createServer()` so don't need the NodeSDK or any monkeypatching. 

This is a first step towards the issues mentioned in this thread, and the proposed `bun-otel` package would provide automatic distributed tracing for Hono, Elysia, and other frameworks.

Basically
```typescript
  import { BunSDK } from 'bun-otel';

const sdk = new BunSDK({ tracerProvider });
sdk.start();

// All requests now get a span
  Bun.serve({ /* ... */ });
```

There is more work to do (e.g. fetch header propagation, `AsyncLocalStorage` for timers, promises, logs, etc) but I'm looking for feedback on the approach before moving to Phase 2 (AsyncLocalStorage integration). Would love to hear thoughts from folks who've been working around this limitation, is this directionally correct? Happy to take comments on [PR #23798](https://github.com/oven-sh/bun/pull/23798)

**Comment by theshadow27** [comment]
The MVP #24063 is probably ready for alpha testing with real exporters / auto instrumentation. It's a huge PR, the more eyes on it the better. ~450 coderabbit comments resolved, ~100 to go. It will get there. 

Basically zero cost when off. To enable reasonable defaults:

```ts
import { BunSDK } from 'bun-otel'
using sdk = new BunSDK();
await sdk.start();

// for the pedantic 
// await sdk.startAndRegisterSystemShutdownHooks();
```

That's it. Distributed tracing, w3c propagation, and metrics reporting on Bun.serve, http.createServer, fetch, and http.request. That covers the new ( @elysiajs , hono) and old (express, NestJS).  

I added basic `sqlite` to show the interface is sufficient beyond http. The design makes it easy to extend incrementally to native redis, s3, sql, etc. in the future - but at +17,000 I'm trying not to let the scope creep without some feedback from the @oven-sh team.

Do let them know if this is important to your adoption of bun! We can't use it in prod without this so really hoping to get the support to get this merged soon!
